File: frontend/src/observationProvenance.ts
import type { Observation } from "./types";

export const LEGACY_UNCONFIRMED_UPLOAD_DEVICE_ID = "upload";
export const BODY_PHOTO_UPLOAD_DEVICE_ID = "manual-upload:body-photo";
export const OTHER_IMAGE_UPLOAD_DEVICE_ID = "manual-upload:other-image";
export const MANUAL_UPLOAD_PREFIX = "manual-upload:";
export const CLINICAL_PHOTO_DEVICE_ID = "clinical-photo:operator-confirmed";

export type UploadSourceType =
  | "microscope"
  | "body_photo"
  | "other_image"
  | "unknown";

export function hasConfirmedMicroscopeSource(
  observation: Pick<Observation, "device_id">,
) {
  return inferUploadSourceType(observation) === "microscope";
}

export function inferUploadSourceType(
  observation: Pick<Observation, "device_id">,
): UploadSourceType {
  if (observation.device_id === LEGACY_UNCONFIRMED_UPLOAD_DEVICE_ID) {
    return "unknown";
  }
  if (observation.device_id === CLINICAL_PHOTO_DEVICE_ID) {
    return "microscope";
  }
  if (observation.device_id === BODY_PHOTO_UPLOAD_DEVICE_ID) {
    return "body_photo";
  }
  if (observation.device_id === OTHER_IMAGE_UPLOAD_DEVICE_ID) {
    return "other_image";
  }
  if (
    observation.device_id.startsWith(MANUAL_UPLOAD_PREFIX) &&
    observation.device_id !== "manual-upload:usb-microscope-confirmed"
  ) {
    return "other_image";
  }
  return "microscope";
}

export function isAIEligibleObservation(observation: Observation) {
  return aiIneligibilityReason(observation) === null;
}

export function aiIneligibilityReason(observation: Observation) {
  if (observation.quality_status !== "accepted") {
    return "The image did not pass the quality gate.";
  }
  if (!hasConfirmedMicroscopeSource(observation)) {
    return "Only an operator-confirmed lesion image can be analyzed.";
  }
  return null;
}

File: backend/app/observation_provenance.py
from .models import Observation

LEGACY_UNCONFIRMED_UPLOAD_DEVICE_ID = "upload"
MANUAL_UPLOAD_SOURCE_PREFIX = "manual-upload:"
CONFIRMED_MANUAL_MICROSCOPE_DEVICE_ID = (
    f"{MANUAL_UPLOAD_SOURCE_PREFIX}usb-microscope-confirmed"
)
CONFIRMED_CLINICAL_PHOTO_DEVICE_ID = "clinical-photo:operator-confirmed"
MANUAL_UPLOAD_BODY_PHOTO_DEVICE_ID = f"{MANUAL_UPLOAD_SOURCE_PREFIX}body-photo"
MANUAL_UPLOAD_OTHER_IMAGE_DEVICE_ID = f"{MANUAL_UPLOAD_SOURCE_PREFIX}other-image"


def is_upload_device_id(device_id: str) -> bool:
    return device_id == LEGACY_UNCONFIRMED_UPLOAD_DEVICE_ID or device_id.startswith(
        MANUAL_UPLOAD_SOURCE_PREFIX
    )


def has_confirmed_microscope_source(observation: Observation) -> bool:
    if observation.device_id == CONFIRMED_CLINICAL_PHOTO_DEVICE_ID:
        return True
    if is_upload_device_id(observation.device_id):
        return observation.device_id == CONFIRMED_MANUAL_MICROSCOPE_DEVICE_ID
    return True


def infer_observation_source_type(device_id: str) -> str:
    if device_id == CONFIRMED_CLINICAL_PHOTO_DEVICE_ID:
        return "clinical_photo_candidate"
    if device_id == LEGACY_UNCONFIRMED_UPLOAD_DEVICE_ID:
        return "legacy_upload"
    if device_id == CONFIRMED_MANUAL_MICROSCOPE_DEVICE_ID:
        return "manual_microscope"
    if device_id == MANUAL_UPLOAD_BODY_PHOTO_DEVICE_ID:
        return "manual_body_photo"
    if device_id == MANUAL_UPLOAD_OTHER_IMAGE_DEVICE_ID:
        return "manual_other_image"
    if is_upload_device_id(device_id):
        return "manual_other_image"
    return "microscope_capture"
